### PR TITLE
Design doc for expanded entity

### PR DESCRIPTION
**Pull request by biaol-odata**
### Issues

*This pull request is to create design doc related to issue #1265 .*

### Description

*Design doc: $expand for context url in V4.01.*

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: v7/_posts/design/2018-09-26-$expandForV4.01.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    "FirstName": "Russell",
+    "LastName": "Whyte"
+}
+```
+As shown above, @odata.context annotation specifies the context url, and provides the following machine-readable description for response data:
+- It is regarding an entity from the "People" entity set,
+- It contains two property values "FirstName", "LastName".
+
+## Expanded Entity Specification for OData V4.01
+In OData V4.01, a different [format](http://docs.oasis-open.org/odata/odata/v4.01/cs01/part1-protocol/odata-v4.01-cs01-part1-protocol.html#sec_ExpandedEntity) is introduced for $expand clause (for expanded entity and projected expanded entity).
```
Comment (mikepizzo) [inline v7/_posts/design/2018-09-26-$expandForV4.01.md]: 4.01 is not really a different format; more accurate would be to say that 4.01 differentiates between a selected navigation property (which appears without parens) and an expanded navigation property (which appears suffixed with parens, even if there are no nested expand options).

File: v7/_posts/design/2018-09-26-$expandForV4.01.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+For a 4.0 response, the expanded navigation property suffixed with parentheses MAY be omitted from the select-list if it does not contain a nested $select or $expand, but MUST still be present, without a suffix, if it is explicitly selected.
+```
+
+## Change Summary
+According to the OData spec above, this change is for creating proper expand token in response's context url, corresponding to $expand clause in the request url. Context Url is used as response metadata and is used to control data materialization when response is received by the client.
+We need to make sure that:
+- Context url in the response generated contains correct token for the $expand clause.
+- Client can parse the context url correctly and uses the metadata for response processing.
+
+### V4.01
```
Comment (mikepizzo) [inline v7/_posts/design/2018-09-26-$expandForV4.01.md]: Is this behavior consistent between 4.0 and 4.01 (desired) or only for 4.01?  Unless we think the empty parenthesis will break clients (I think we determined it did not for .NET client, other than a related bug that treated nav props wrong in general) we should just always return the parenthesized nav prop when expanded (not dependent on response version).

File: v7/_posts/design/2018-09-26-$expandForV4.01.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+As required, we are going to add to the context url the parenthesized (either empty or non-empty) navigation property when it is expanded. For example:
+
+Request Url | Context Url in Response
+----------- | -----------------------
+ root/Cities('id')?$expand=TestModel.CapitolCity/Districts                       | root/$metadata#Cities(TestModel.CapitolCity/Districts())/$entity
+ root/Cities('id')?$expand=TestModel.CapitolCity/Districts($select=DistrictName) | root/$metadata#Cities(TestModel.CapitolCity/Districts(DistrictName))/$entity
+ root/Cities('id')?$select=Name&$expand=TestModel.CapitolCity/Districts($select=DistrictName) | root/$metadata#Cities(Name,TestModel.CapitolCity/Districts(DistrictName))/$entity
+ root/Cities('id')?$select=Name,TestModel.CapitolCity/Districts$expand=TestModel.CapitolCity/Districts($select=DistrictName) | root/$metadata#Cities(Name,TestModel.CapitolCity/Districts(DistrictName))/$entity
```
Comment (mikepizzo) [inline v7/_posts/design/2018-09-26-$expandForV4.01.md]: >root/$metadata#Cities(Name,TestModel.CapitolCity/Districts(DistrictName))/$entity [](start = 127, length = 81)

shouldn't this be: root/$metadata#Cities(Name,TestModel.CapitolCity/Districts,TestModel.CapitolCity/Districts(DistrictName))/$entity ?

File: v7/_posts/design/2018-09-26-$expandForV4.01.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+
+Except for the case (due to known issue https://github.com/OData/odata.net/issues/1104) when a navigation link is both expanded and selected, the parenthesized navigation property will not be present in the context-url, which only has the navigation link in the selected list:
+
```
Comment (mikepizzo) [inline v7/_posts/design/2018-09-26-$expandForV4.01.md]: This is current behavior or new behavior? New behavior should be, if a nav prop is both selected and expanded, we should have both the parenthesized nav prop and the unparenthesized nav prop.

File: v7/_posts/design/2018-09-26-$expandForV4.01.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+Request Url | Context Url in Response
+----------- | -----------------------
+ root/Cities('id')?$select=**TestModel.CapitolCity/Districts**&$expand=**TestModel.CapitolCity/Districts** | root/$metadata#Cities(**TestModel.CapitolCity/Districts**)/$entity
```
Comment (mikepizzo) [inline v7/_posts/design/2018-09-26-$expandForV4.01.md]: >root/$metadata#Cities(**TestModel.CapitolCity/Districts**)/$entity [](start = 109, length = 66)

After code change, should be root/$metadata#Cities(TestModel.CapitolCity/Districts,TestModel.CapitolCity/Districts())/$entity

File: v7/_posts/design/2018-09-26-$expandForV4.01.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+ root/Cities('id')?$select=Id,Name,**ExpandedNavProp**&$expand=**ExpandedNavProp**                         | root/$metadata#Cities(Id,Name,**ExpandedNavProp**)/$entity
```
Comment (mikepizzo) [inline v7/_posts/design/2018-09-26-$expandForV4.01.md]: >root/$metadata#Cities(Id,Name,**ExpandedNavProp**)/$entity [](start = 109, length = 58)

after code change should be root/$metadata#Cities(Id,Name,ExpandedNavProp,ExpandedNavProp())/$entity

File: v7/_posts/design/2018-09-26-$expandForV4.01.md (hunk 4 above)
Comment (biaol-odata) [reply v7/_posts/design/2018-09-26-$expandForV4.01.md]: This is current behavior due to issue #1104. If we can fix it, there would be no exceptions. However, trying to fix this leaded to issue #1268, which requires changes regarding representation of expand nav properties in SelectedPropertiesNode. We should find some time to go over the approach in my mind.

Comment (biaol-odata) [reply v7/_posts/design/2018-09-26-$expandForV4.01.md]: We should be able to expand the scope of this task to fix issue #1104, to make selected navigation properties really explicit.
Similar comments below are related to fixing #1104, same resolution is applicable.

File: v7/_posts/design/2018-09-26-$expandForV4.01.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    "FirstName": "Russell",
+    "LastName": "Whyte"
+}
+```
+As shown above, @odata.context annotation specifies the context url, and provides the following machine-readable description for response data:
+- It is regarding an entity from the "People" entity set,
+- It contains two property values "FirstName", "LastName".
+
+## Expanded Entity Specification for OData V4.01
+In OData V4.01, notion for expanded entity and projected expanded entity [format](http://docs.oasis-open.org/odata/odata/v4.01/cs01/part1-protocol/odata-v4.01-cs01-part1-protocol.html#sec_ExpandedEntity) is introduced for $expand clause in context url.
```
Comment (mikepizzo) [inline v7/_posts/design/2018-09-26-$expandForV4.01.md]: Note that expanded nav props with nested selects were always included in the contextUrl.  The only change in 4.01 was that expanded nav props with no nested selects/expands are included with empty parens.  So more accurate would be:

OData V4.0 specifies that expanded entities with nested selects are included in the context Url as the name of the navigation property suffixed with the comma separated list of selected properties, enclosed in parens. OData V4.01 [format](http://docs.oasis-open.org/odata/odata/v4.01/cs01/part1-protocol/odata-v4.01-cs01-part1-protocol.html#sec_ExpandedEntity) specifies that, in the absence of any nested selects, the expanded navigation property appears suffixed with empty parens. This is distinct, and may appear in addition to, the un-suffixed navigation property name, which indicates that the navigation property appears in the $select list (indicating that the navigationLink should be included in the response).

File: v7/_posts/design/2018-09-26-$expandForV4.01.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+We also fix the known issue https://github.com/OData/odata.net/issues/1104 when a navigation link is both expanded and selected, the parenthesized navigation property will be present in the context-url, along with the navigation link explicitly selected:
+
+Request Url | Context Url in Response
+----------- | -----------------------
+ root/Cities('id')?$select=**TestModel.CapitolCity/Districts**&$expand=**TestModel.CapitolCity/Districts** | root/$metadata#Cities(**TestModel.CapitolCity/Districts,TestModel.CapitolCity/Districts()**)/$entity
+ root/Cities('id')?$select=Id,Name,**ExpandedNavProp**&$expand=**ExpandedNavProp**                         | root/$metadata#Cities(Id,Name,**ExpandedNavProp,ExpandedNavProp()**)/$entity
+
+### V4.0
+We are going to tweet the behavior to align with that of the 4.01. While this change does introduce slightly different semantics for explicitly selected navigation property, it is not considered a breaking change because the updated context-url form for expanded navigation link is legal, but just was not required. It has been confirmed through code examination and ODL tests that this change doesn't cause anomalies for libraries and client.
```
Comment (mikepizzo) [inline v7/_posts/design/2018-09-26-$expandForV4.01.md]: >tweet [](start = 16, length = 5)

I think you mean "tweak"

File: v7/_posts/design/2018-09-26-$expandForV4.01.md (hunk 7 above)
Comment (biaol-odata) [reply v7/_posts/design/2018-09-26-$expandForV4.01.md]: Updated as suggested. Thanks.

---
In reply to: [227443781](https://github.com/OData/odata.net/pull/1266#discussion_r227443781) [](ancestors = 227443781)

**Pull request merged**